pipelines/templates/api-deploy-environment.yml

```diff
@@ -0,0 +1,94 @@
+    displayName: Deploy API Backend
+    strategy:
+      runOnce:
+        deploy:
+          steps:
+            - bash: echo "Deploying ${{ parameters.environment }} environment"
+              displayName: Check for Approval
+  - job: setup
+    displayName: Initialize
+    variables:
+      - name: regionList
+        value: ${{ join(',', parameters.regions) }}
+    steps:
+      - task: Bash@3
+        name: createMatrixJson
+        displayName: Create Matrix
+        inputs:
+          targetType: inline
+          script: |
+            echo $(regionList)
+            source ./deployment/arm/utils.sh
+            matrix=$(toRegionMatrix $(regionList))
+            echo $matrix
+            echo "##vso[task.setVariable variable=matrix;isOutput=true]$matrix"
+
+  - job: RegionalMaxtrixDeployment
+    displayName: "Deploy Region:"
+    dependsOn: setup
+    strategy:
+      matrix: $[ dependencies.setup.outputs['createMatrixJson.matrix'] ]
+    steps:
+      - task: UseDotNet@2
+        displayName: Dotnet version 3.1.100
+        inputs:
+          version: '3.1.100'
+
+      - task: DownloadBuildArtifacts@0
+        displayName: Download Build Artifacts
+        inputs:
+          buildType: 'current'
+          downloadType: 'single'
+          artifactName: 'drop'
+          downloadPath: '$(Build.ArtifactStagingDirectory)'
+
+      - task: AzureRMWebAppDeployment@4
+        displayName: Deploy to Staging Slot
+        inputs:
+          ConnectionType: 'AzureRM'
+          azureSubscription: ${{ parameters.subscription }}
+          appType: 'apiApp'
+          ResourceGroupName: 'rg-${{ parameters.businessUnit }}-${{ parameters.appName }}-${{ parameters.environment }}-$(region)'
+          WebAppName: 'backend-${{ parameters.businessUnit }}-${{ parameters.appName }}-${{ parameters.environment }}-$(region)'
+          deployToSlotOrASE: true
+          SlotName: staging
+          packageForLinux: '$(Build.ArtifactStagingDirectory)/**/*.zip'  
+          appSettings: 
+            -CosmosSettings:ServiceEndpoint $(COSMOSDB_ENDPOINT) 
+            -CosmosSettings:AuthKey $(COSMOSDB_AUTHKEY) 
+            -CosmosSettings:DatabaseName linkylinkdb 
+            -CosmosSettings:ContainerName linkbundles
+            -AzureAdB2C:Instance $(AZUREADB2C_INSTANCE)
+            -AzureAdB2C:ClientId $(AZUREADB2C_CLIENTID)
+            -AzureAdB2C:Domain $(AZUREADB2C_DOMAIN)
+            -AzureAdB2C:SignUpSignInPolicyId $(AZUREADB2C_POLICYID)
+
+      - task: DotNetCoreCLI@2
+        displayName: Validation Tests
+        inputs:
+          command: test
+          projects: '**/LinkyLink.Integration.Tests/*.csproj'
+          arguments: -c $(BuildConfiguration)
+        env:
+          INTTEST_AzureAdB2C__BaseAddress: 'https://backend-${{ parameters.businessUnit }}-${{ parameters.appName }}-${{ parameters.environment }}-$(region)-staging.azurewebsites.net'
+      
+      - task: AzureAppServiceManage@0
```

Comment:
Can we guard this task with a human gate / interaction? Otherwise, is there really a benefit for deploying to staging and swapping?

Reply:
I agree with having some gate-check. But regarding the human part, is this something that you would like to utilize in the challenge Andy? If not, we can start with it and then, if we have time, can we add an automated gate that would check failure-rate within a given period of time (assuming we have the KPI enabled from the get-go), if there are no failures/alarms then continue with the deployment to prod? 

Man! I hate it when the comments do not get updated with latest code location!

Reply:
Right now we have the manual gate to production.  We are using the same deployment scripts/pipelines across all environments so the deploy to staging slot & swap is done everywhere.  I actually like this since if an integration tests fails anywhere it will stop any additional deployments.